core/selection.js

```diff
@@ -943,10 +943,24 @@
 				// https://dev.ckeditor.com/ticket/14407 - Don't even let anything happen if the selection is in a non-editable element.
 				editable.attachListener( editable, 'keydown', function( evt ) {
 					var sel = this.getSelection( 1 ),
-						ascendant = getNonEditableAscendant( sel );
+						ascendant = getNonEditableAscendant( sel ),
+						activeElement,
+						isValidActiveElement;
+
+					if ( !ascendant ) {
+						return;
+					}
+
+					activeElement = this.document.getActive(),
+					isValidActiveElement = activeElement && ( activeElement.getName() === 'input' || activeElement.getName() === 'textarea' );
+
+					// Allow on typing inside editable elements of widgets if those are currently focused (#3587)
+					if ( isValidActiveElement && ascendant.contains( activeElement ) ) {
+						return;
+					}
```

Comment:
Could we improve code readability here by keeping a single abstraction level? I suggest extracting this code into helper function, something like:  `isTypingElement` - function checking if element is of `input`, `textarea` or `contenteditable=true` type.

Also, it seems like in that case, checking for `nonEditableAscendant` could be moved lower, as if an active element is directly editable, there is no need to traverse through parents.